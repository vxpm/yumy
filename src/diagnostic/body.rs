use super::{config::Config, Label};
use crate::{
    source::{Source, SourceLine},
    SourceSpan,
};
use either::Either;
use owo_colors::{OwoColorize, Style};
use std::{io::Write, ops::Range};

#[derive(Debug, Clone, Copy)]
struct IdentInfo {
    end: usize,
    len: usize,
}

fn ident_info(text: &str) -> IdentInfo {
    let mut len = 0;
    let mut c_indices = text.char_indices();

    let end = loop {
        let Some((start, c)) = c_indices.next() else {
            break text.len();
        };

        len += match c {
            ' ' => 1,
            '\t' => 4,
            _ => break start,
        }
    };

    IdentInfo { end, len }
}

#[derive(Debug, Clone)]
struct SinglelineLabel {
    message: String,
    line: u32,
    line_span: SourceSpan,
    indicator_style: Option<Style>,
}

#[derive(Debug, Clone)]
struct MultilineLabel {
    message: String,
    line_range: Range<u32>,
    indicator_style: Option<Style>,
}

#[derive(Debug, Clone, Default)]
enum Slot {
    RecentlyAdded(MultilineLabel),
    Active(MultilineLabel),
    #[default]
    Inactive,
}

impl Slot {
    pub fn is_active(&self) -> bool {
        match self {
            Slot::RecentlyAdded(_) => true,
            Slot::Active(_) => true,
            Slot::Inactive => false,
        }
    }

    pub fn unwrap_label(self) -> MultilineLabel {
        match self {
            Slot::RecentlyAdded(label) => label,
            Slot::Active(label) => label,
            Slot::Inactive => panic!("tried to call unwrap_label on inactive slot"),
        }
    }
}

/// Struct that takes care of emitting the body of a diagnostic.
/// Keeping the state for this in it's own struct is easier.
pub(crate) struct BodyWriter<'src, W> {
    writer: W,
    source: Source<'src>,
    config: Config,
    left_padding: usize,
    ident_len: usize,
    singleline_labels: Vec<SinglelineLabel>,
    multiline_labels: Vec<MultilineLabel>,
    multiline_slots: Vec<Slot>,
    current_line: u32,
}

impl<'src, W> BodyWriter<'src, W>
where
    W: Write,
{
    /// Calculates the number of slots needed for a set
    /// of multiline labels.
    fn slots_needed(labels: &[MultilineLabel]) -> usize {
        enum Event {
            Start(u32),
            End(u32),
        }

        impl Event {
            fn index(&self) -> u32 {
                match self {
                    Event::Start(x) => *x,
                    Event::End(x) => *x,
                }
            }
        }

        let mut events: Vec<_> = labels
            .iter()
            .flat_map(|x| {
                [
                    Event::Start(x.line_range.start),
                    Event::End(x.line_range.end),
                ]
            })
            .collect();

        events.sort_unstable_by(|a, b| a.index().cmp(&b.index()));

        let mut current = 0;
        let mut max = 0;

        for event in events {
            match event {
                Event::Start(_) => current += 1,
                Event::End(_) => current -= 1,
            }

            if current > max {
                max = current;
            }
        }

        max
    }

    pub(crate) fn new(
        writer: W,
        source: Source<'src>,
        config: &Config,
        left_padding: usize,
        labels: &[Label],
    ) -> Self {
        let mut singleline_labels = Vec::new();
        let mut multiline_labels = Vec::new();
        for label in labels {
            if label.is_singleline(&source) {
                let line_index = label.line_range(&source).start;
                let line = source.line(line_index).unwrap();

                let label_line_start = label.span.start() - line.span.start();
                let label_line_end = label_line_start + label.span.len();

                let label = SinglelineLabel {
                    message: label.message.clone(),
                    line: label.line_range(&source).start,
                    line_span: SourceSpan::new(label_line_start, label_line_end),
                    indicator_style: label.indicator_style,
                };
                singleline_labels.push(label);
            } else {
                let label = MultilineLabel {
                    message: label.message.clone(),
                    line_range: label.line_range(&source),
                    indicator_style: label.indicator_style,
                };

                multiline_labels.push(label);
            }
        }

        let singleline_lines = singleline_labels
            .iter()
            .map(|label| source.line(label.line).unwrap());
        let multiline_lines = multiline_labels.iter().flat_map(|label| {
            [
                source.line(label.line_range.start).unwrap(),
                source.line(label.line_range.end - 1).unwrap(),
            ]
        });

        let ident_width = singleline_lines
            .chain(multiline_lines)
            .map(|line| ident_info(line.line).len)
            .min()
            .unwrap_or(0);

        Self {
            writer,
            source,
            config: config.clone(),
            left_padding,
            ident_len: ident_width,
            singleline_labels,
            multiline_slots: vec![Slot::Inactive; Self::slots_needed(&multiline_labels)],
            multiline_labels,
            current_line: 0,
        }
    }

    /// Returns the next label, be it single or multi line.
    fn next_label(&mut self) -> Option<Either<SinglelineLabel, MultilineLabel>> {
        let next_singleline_label = self
            .singleline_labels
            .iter()
            .map(|x| x.line)
            .enumerate()
            .min_by_key(|(_, start)| *start);

        let next_multiline_label = self
            .multiline_labels
            .iter()
            .map(|x| x.line_range.start)
            .enumerate()
            .min_by_key(|(_, start)| *start);

        let (index, is_singleline) = match (next_singleline_label, next_multiline_label) {
            (None, None) => return None,
            (Some((singleline_index, _)), None) => (singleline_index, true),
            (Some((singleline_index, singleline_start)), Some((_, multiline_start)))
                if singleline_start < multiline_start =>
            {
                (singleline_index, true)
            }
            (_, Some((multiline_index, _))) => (multiline_index, false),
        };

        Some(if is_singleline {
            Either::Left(self.singleline_labels.remove(index))
        } else {
            Either::Right(self.multiline_labels.remove(index))
        })
    }

    /// Returns the next singleline label in the current line.
    fn next_singleline_label_in_current(&mut self) -> Option<SinglelineLabel> {
        self.singleline_labels
            .iter()
            .enumerate()
            .find(|(_, label)| label.line == self.current_line)
            .map(|(index, _)| index)
            .map(|index| self.singleline_labels.remove(index))
    }

    /// Returns the next multiline label starting in the current line.
    fn next_multiline_label_in_current(&mut self) -> Option<MultilineLabel> {
        self.multiline_labels
            .iter()
            .enumerate()
            .find(|(_, label)| label.line_range.start == self.current_line)
            .map(|(index, _)| index)
            .map(|index| self.multiline_labels.remove(index))
    }

    /// Are there any active multiline labels?
    #[inline]
    fn has_active_multiline_labels(&self) -> bool {
        self.multiline_slots.iter().any(|x| x.is_active())
    }

    /// Allocate the given multiline label into an available slot.
    #[inline]
    fn allocate_multiline_label(&mut self, label: MultilineLabel) {
        let slot_index = self
            .multiline_slots
            .iter()
            .enumerate()
            .find(|(_, slot)| !slot.is_active())
            .map(|(index, _)| index);

        if let Some(slot_index) = slot_index {
            self.multiline_slots[slot_index] = Slot::RecentlyAdded(label);
        } else {
            unreachable!("should have enough slots")
        }
    }

    /// Emit the left column of the body.
    #[inline]
    fn emit_left_column(&mut self, line_index: impl Into<Option<usize>>) -> std::io::Result<()> {
        if let Some(index) = line_index.into() {
            write!(
                self.writer,
                "{:padding$} {} ",
                (index + 1).style(self.config.styles.left_column),
                self.config
                    .charset
                    .vertical_bar
                    .style(self.config.styles.left_column),
                padding = self.left_padding
            )?;
        } else {
            write!(
                self.writer,
                "{:padding$} {} ",
                "",
                self.config
                    .charset
                    .separator
                    .style(self.config.styles.left_column),
                padding = self.left_padding
            )?;
        }

        Ok(())
    }

    fn emit_multiline_indicators(&mut self) -> std::io::Result<()> {
        for slot in self.multiline_slots.iter_mut() {
            let (is_new, label) = match slot {
                slot @ Slot::RecentlyAdded(_) => {
                    let Slot::RecentlyAdded(label) = std::mem::take(slot) else {
                        unreachable!()
                    };

                    *slot = Slot::Active(label);
                    let Slot::Active(ref label) = slot else {
                        unreachable!()
                    };

                    (true, label)
                }
                Slot::Active(label) => (false, &*label),
                Slot::Inactive => {
                    write!(self.writer, " ")?;
                    continue;
                }
            };

            let style = label
                .indicator_style
                .unwrap_or(self.config.styles.multiline_indicator);

            let indicator_char = if is_new {
                self.config.charset.multiline_start
            } else if label.line_range.end == self.current_line + 1 {
                self.config.charset.multiline_end
            } else {
                self.config.charset.vertical_bar
            };

            write!(self.writer, "{}", indicator_char.style(style))?;
        }

        Ok(())
    }

    /// Emit the given source line.
    fn emit_source_line(&mut self, line: SourceLine, line_index: u32) -> std::io::Result<()> {
        self.emit_left_column(line_index as usize)?;
        self.emit_multiline_indicators()?;

        let line_ident_info = ident_info(line.line);
        let spaces = line_ident_info.len - self.ident_len;

        let style = self.source.style().unwrap_or(self.config.styles.source);

        write!(self.writer, "{:x$} ", "", x = spaces)?;
        writeln!(
            self.writer,
            "{}",
            (&line.line[line_ident_info.end..]).style(style),
        )?;
        Ok(())
    }

    /// Emit the given singleline label.
    fn emit_singleline_label(
        &mut self,
        line: SourceLine,
        label: SinglelineLabel,
    ) -> std::io::Result<()> {
        self.emit_left_column(None)?;
        self.emit_multiline_indicators()?;

        let line_ident_info = ident_info(line.line);
        let spaces = line_ident_info.len - self.ident_len;

        let line_width = unicode_width::UnicodeWidthStr::width(line.line);

        let before_underliner_width = spaces
            + unicode_width::UnicodeWidthStr::width(&line.line[..label.line_span.start() as usize]);
        let after_underliner_width =
            unicode_width::UnicodeWidthStr::width(&line.line[label.line_span.end() as usize..]);
        let underliner_width = line_width - (before_underliner_width + after_underliner_width);

        let before = std::iter::repeat(' ').take(before_underliner_width);
        let underliner = std::iter::repeat(self.config.charset.underliner).take(underliner_width);

        let underliner = before.chain(underliner);

        let style = label
            .indicator_style
            .unwrap_or(self.config.styles.singleline_indicator);

        write!(self.writer, "{:x$}", "", x = spaces)?;
        for c in underliner {
            write!(self.writer, "{}", c.style(style))?;
        }

        writeln!(self.writer, " {}", label.message)?;
        Ok(())
    }

    /// Emits all singleline labels in the current line.
    #[inline]
    fn emit_singleline_labels_in_current(&mut self, line: SourceLine) -> std::io::Result<()> {
        while let Some(label) = self.next_singleline_label_in_current() {
            self.emit_singleline_label(line, label)?;
        }

        Ok(())
    }

    /// Emit the end of the given multiline label.
    fn emit_multiline_label_end(
        &mut self,
        line: SourceLine,
        label: MultilineLabel,
        label_slot: u32,
    ) -> std::io::Result<()> {
        self.emit_left_column(None)?;
        let line_width = unicode_width::UnicodeWidthStr::width(line.line);
        let this_style = label.indicator_style;

        for slot in &self.multiline_slots[..label_slot as usize] {
            match slot {
                Slot::RecentlyAdded(_) => {
                    unreachable!("singleline multiline labels should be impossible")
                }
                Slot::Active(label) => {
                    let style = label
                        .indicator_style
                        .unwrap_or(self.config.styles.multiline_indicator);

                    write!(
                        self.writer,
                        "{}",
                        self.config.charset.vertical_bar.style(style)
                    )?;
                }
                Slot::Inactive => {
                    write!(self.writer, " ")?;
                }
            }
        }

        write!(
            self.writer,
            "{}",
            self.config
                .charset
                .connection_top_to_right
                .style(this_style.unwrap_or(self.config.styles.multiline_indicator))
        )?;

        for slot in &self.multiline_slots[label_slot as usize + 1..] {
            match slot {
                Slot::RecentlyAdded(_) => {
                    unreachable!("singleline multiline labels should be impossible")
                }
                Slot::Active(label) => {
                    let style = this_style.unwrap_or(
                        label
                            .indicator_style
                            .unwrap_or(self.config.styles.multiline_indicator),
                    );

                    write!(
                        self.writer,
                        "{}",
                        self.config.charset.multiline_crossing.style(style)
                    )?;
                }
                Slot::Inactive => {
                    write!(
                        self.writer,
                        "{}",
                        self.config
                            .charset
                            .horizontal_bar
                            .style(this_style.unwrap_or(self.config.styles.multiline_indicator))
                    )?;
                }
            }
        }

        let underliner: std::iter::Take<std::iter::Repeat<char>> =
            std::iter::repeat(self.config.charset.horizontal_bar).take(line_width + 1);
        for c in underliner {
            write!(
                self.writer,
                "{}",
                c.style(this_style.unwrap_or(self.config.styles.multiline_indicator))
            )?;
        }

        writeln!(self.writer, " {}", label.message)?;
        Ok(())
    }

    /// Try to finish any active multiline labels that end in the current line.
    fn try_finishing_active_multiline_labels(&mut self) -> std::io::Result<()> {
        let mut multiline_slot = 0;
        while multiline_slot < self.multiline_slots.len() {
            let Slot::Active(label) = &self.multiline_slots[multiline_slot] else {
                multiline_slot += 1;
                continue;
            };

            if label.line_range.end == self.current_line + 1 {
                let label =
                    std::mem::take(&mut self.multiline_slots[multiline_slot]).unwrap_label();

                let line = self.source.line(self.current_line).unwrap();
                self.emit_multiline_label_end(line, label, multiline_slot as u32)?;
            } else {
                multiline_slot += 1;
            }
        }

        Ok(())
    }

    /// Allocate all multiline labels that start in the current line.
    #[inline]
    fn start_new_multiline_labels(&mut self) {
        while let Some(label) = self.next_multiline_label_in_current() {
            self.allocate_multiline_label(label);
        }
    }

    pub(crate) fn write(mut self) -> std::io::Result<()> {
        // sort singleline labels from biggest to smallest
        self.singleline_labels
            .sort_unstable_by_key(|x| std::cmp::Reverse(x.line_span.len()));

        // sort multiline labels from bottom to top (relative to the end)
        self.multiline_labels
            .sort_unstable_by_key(|x| std::cmp::Reverse(x.line_range.end));

        loop {
            if !self.has_active_multiline_labels() {
                let Some(label) = self.next_label() else {
                    break;
                };

                match label {
                    Either::Left(label) => {
                        let line_index = label.line;
                        self.current_line = line_index;

                        let line = self.source.line(line_index).unwrap();
                        self.emit_source_line(line, line_index)?;
                        self.emit_singleline_label(line, label)?;

                        self.emit_singleline_labels_in_current(line)?;
                    }
                    Either::Right(label) => {
                        let line_index = label.line_range.start;
                        self.current_line = line_index;

                        self.allocate_multiline_label(label);
                        self.start_new_multiline_labels();
                    }
                }
            } else {
                self.start_new_multiline_labels();

                let line = self.source.line(self.current_line).unwrap();
                self.emit_source_line(line, self.current_line)?;

                self.emit_singleline_labels_in_current(line)?;
                self.try_finishing_active_multiline_labels()?;

                self.current_line += 1;
            }
        }

        Ok(())
    }
}
