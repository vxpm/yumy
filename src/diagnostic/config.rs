use owo_colors::Style;

/// The charset to use when rendering a diagnostic.
#[derive(Debug, Clone)]
pub struct Charset {
    /// A vertical bar.
    pub vertical_bar: char,
    /// An horizontal bar.
    pub horizontal_bar: char,
    /// The character used to underline the source
    /// in single-line labels.
    pub underliner: char,
    /// The character that's used instead of the vertical
    /// bar when not in a source line.
    pub separator: char,
    /// The character that connects the vertical bar
    /// to the connector in multiline labels.
    pub connection_top_to_right: char,
    /// The character for when a multiline label starts.
    pub multiline_start: char,
    /// The character for when a multiline label ends.
    pub multiline_end: char,
    /// The character for when two multiline labels cross.
    pub multiline_crossing: char,
}

impl Default for Charset {
    fn default() -> Self {
        Self {
            vertical_bar: '│',
            horizontal_bar: '╶',
            underliner: '^',
            separator: ':',
            connection_top_to_right: '╰',
            multiline_start: '┬',
            multiline_end: '┼',
            multiline_crossing: '┼',
        }
    }
}

/// Default styles to use for each part of a diagnostic.
#[derive(Debug, Clone)]
pub struct DefaultStyles {
    pub source_name: Style,
    pub source: Style,
    pub left_column: Style,
    pub multiline_indicator: Style,
    pub singleline_indicator: Style,
    pub footnote_indicator: Style,
}

impl Default for DefaultStyles {
    fn default() -> Self {
        Self {
            source_name: Style::new().white().bold(),
            source: Style::new().white(),
            left_column: Style::new().bright_blue().bold(),
            multiline_indicator: Style::new().yellow(),
            singleline_indicator: Style::new().yellow(),
            footnote_indicator: Style::new().bright_blue().bold(),
        }
    }
}

/// Configuration used to render a diagnostic.
#[derive(Debug, Clone, Default)]
pub struct Config {
    pub charset: Charset,
    pub styles: DefaultStyles,
}
