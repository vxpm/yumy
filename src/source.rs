use nonmax::NonMaxU32;
use owo_colors::Style;

/// Unit struct that represents the absence of
/// a source in a diagnostic.
#[derive(Debug, Clone, Copy)]
pub struct NoSource;

/// A span into the source of a [`crate::Diagnostic`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct SourceSpan {
    start: NonMaxU32,
    end: u32,
}

impl SourceSpan {
    /// Creates a new source span. `start` and `end`
    /// are byte indexes into the source.
    pub fn new(start: u32, end: u32) -> Self {
        assert!(end >= start);
        Self {
            start: NonMaxU32::new(start).expect("Start is non-max"),
            end,
        }
    }

    /// The start of this span. Inclusive.
    #[inline]
    pub fn start(&self) -> u32 {
        self.start.get()
    }

    /// The end of this span. Exclusive.
    #[inline]
    pub fn end(&self) -> u32 {
        self.end
    }

    /// The length of this span.
    #[inline]
    pub fn len(&self) -> u32 {
        self.end - self.start()
    }

    /// Does this span contain the given value?
    #[inline]
    pub fn contains(&self, value: u32) -> bool {
        (self.start() <= value) && (value < self.end)
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct SourceLine<'src> {
    pub span: SourceSpan,
    pub line: &'src str,
}

impl<'src> SourceLine<'src> {
    pub fn new(line: &'src str, span: SourceSpan) -> Self {
        Self { span, line }
    }
}

/// A source of text to use with a diagnostic.
#[derive(Debug, Clone)]
pub struct Source<'src> {
    src: &'src str,
    name: Option<&'src str>,
    style: Option<Style>,
    lines: Vec<SourceLine<'src>>,
}

impl<'src> Source<'src> {
    fn lines_of(src: &str) -> Vec<SourceLine> {
        let base_addr = src.as_ptr();
        let lines = src.lines().map(|line| {
            let line_addr = line.as_ptr();
            let offset = (line_addr as usize)
                .checked_sub(base_addr as usize)
                .expect("line should always have higher address");
            let end = offset + line.len();

            SourceLine {
                span: SourceSpan::new(offset as u32, end as u32),
                line,
            }
        });

        lines.collect()
    }

    /// Creates a new source.
    pub fn new(src: &'src str, name: Option<&'src str>) -> Self {
        Self {
            src,
            name,
            style: None,
            lines: Self::lines_of(src),
        }
    }

    /// Creates a new source with the given style.
    pub fn styled(src: &'src str, name: Option<&'src str>, style: Style) -> Self {
        Self {
            src,
            name,
            style: Some(style),
            lines: Self::lines_of(src),
        }
    }

    pub fn src(&self) -> &'src str {
        self.src
    }

    pub fn name(&self) -> Option<&'src str> {
        self.name
    }

    pub fn style(&self) -> Option<Style> {
        self.style
    }

    pub fn line(&self, index: u32) -> Option<SourceLine<'src>> {
        self.lines.get(index as usize).copied()
    }

    pub fn lines(&self) -> impl Iterator<Item = &SourceLine> {
        self.lines.iter()
    }

    pub(crate) fn line_index_of_byte(&self, byte_index: u32) -> Option<u32> {
        if byte_index > self.src.len() as u32 {
            return None;
        }

        // TODO: change to a binary search
        for (index, line) in self.lines.iter().enumerate() {
            if line.span.contains(byte_index) {
                return Some(index as u32);
            }

            if line.span.start() > byte_index {
                return Some((index - 1) as u32);
            }
        }

        None
    }
}

#[cfg(test)]
mod test {
    use super::*;

    const SAMPLE: &str = include_str!("../sample.txt");

    #[test]
    fn test_lines() {
        let src = Source::new(SAMPLE, None);
        let mut lines = src.lines();

        assert_eq!(
            Some(&SourceLine {
                span: SourceSpan::new(0, 20),
                line: "hello there darling!"
            }),
            lines.next()
        );

        // notice how index 20 is not included in any line
        // since it is the \n character!

        assert_eq!(
            Some(&SourceLine {
                span: SourceSpan::new(21, 28),
                line: "this is"
            }),
            lines.next()
        );

        // the same goes for 28

        assert_eq!(
            Some(&SourceLine {
                span: SourceSpan::new(29, 29),
                line: ""
            }),
            lines.next()
        );

        // index 29 is the \n! but it is also not included since
        // the line is empty (29..29)

        assert_eq!(
            Some(&SourceLine {
                span: SourceSpan::new(30, 46),
                line: "a sample text :)"
            }),
            lines.next()
        );
    }
}
