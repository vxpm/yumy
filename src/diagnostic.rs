mod body;

/// Module for diagnostic configuration related items.
pub mod config;

use self::config::Config;
use super::source::{NoSource, Source, SourceSpan};
use body::BodyWriter;
use owo_colors::{OwoColorize, Style};
use std::{
    io::{BufWriter, Write},
    ops::Range,
};

/// A label is a message that points to a specific
/// part of the source of a [`Diagnostic`].
#[derive(Debug, Clone)]
pub struct Label {
    /// The message of this label.
    pub message: String,
    /// The span this label refers to.
    pub span: SourceSpan,
    /// The indicator style of this label.
    pub indicator_style: Option<Style>,
}

impl Label {
    /// Creates a new label.
    pub fn new<S, M>(span: S, message: M) -> Self
    where
        S: Into<SourceSpan>,
        M: ToString,
    {
        Self {
            message: message.to_string(),
            span: span.into(),
            indicator_style: None,
        }
    }

    /// Creates a new label with the given style for it's indicator.
    pub fn styled<S, M>(span: S, message: M, style: Style) -> Self
    where
        S: Into<SourceSpan>,
        M: ToString,
    {
        Self {
            message: message.to_string(),
            span: span.into(),
            indicator_style: Some(style),
        }
    }

    /// Returns the line range of this label in the given source.
    ///
    /// # Panics
    /// Panics if the span is out of bounds.
    fn line_range(&self, src: &Source) -> Range<u32> {
        src.line_range_of_span(self.span)
            .expect("label should have span in range")
    }

    /// Returns whether this label is singleline or not.
    ///
    /// # Panics
    /// Panics if [`Label::line_range`] would panic.
    fn is_singleline(&self, src: &Source) -> bool {
        let line_range = self.line_range(src);
        line_range.start + 1 == line_range.end
    }
}

/// A diagnostic.
#[derive(Debug, Clone)]
pub struct Diagnostic<Src> {
    message: String,
    labels: Vec<Label>,
    footnotes: Vec<String>,
    source: Src,
}

impl Diagnostic<NoSource> {
    /// Create a new diagnostic without an associated source.
    #[inline]
    pub fn new<M>(message: M) -> Self
    where
        M: ToString,
    {
        Self {
            message: message.to_string(),
            labels: Vec::new(),
            footnotes: Vec::new(),
            source: NoSource,
        }
    }

    /// Attach a source to this diagnostic.
    #[inline(always)]
    pub fn with_source(self, source: Source<'_>) -> Diagnostic<Source<'_>> {
        Diagnostic {
            message: self.message,
            labels: self.labels,
            footnotes: self.footnotes,
            source,
        }
    }
}

impl<Src> Diagnostic<Src> {
    /// Add a [`Label`] to this diagnostic.
    #[inline(always)]
    pub fn with_message<M>(mut self, message: M) -> Self
    where
        M: ToString,
    {
        self.message = message.to_string();
        self
    }

    /// Add a [`Label`] to this diagnostic.
    #[inline(always)]
    pub fn add_label(&mut self, label: Label) {
        self.labels.push(label);
    }

    /// Add a [`Label`] to this diagnostic.
    #[inline(always)]
    pub fn with_label(mut self, label: Label) -> Self {
        self.add_label(label);
        self
    }

    /// Replaces the [`Label`]s of this diagnostic.
    #[inline(always)]
    pub fn with_labels(mut self, labels: Vec<Label>) -> Self {
        self.labels = labels;
        self
    }

    /// Add a footnote to this diagnostic. A footnote is a message
    /// shown after the body of a diagnostic.
    #[inline(always)]
    pub fn add_footnote<F>(&mut self, footnote: F)
    where
        F: ToString,
    {
        self.footnotes.push(footnote.to_string());
    }

    /// Add a footnote to this diagnostic. A footnote is a message
    /// shown after the body of a diagnostic.
    #[inline(always)]
    pub fn with_footnote<F>(mut self, footnote: F) -> Self
    where
        F: ToString,
    {
        self.add_footnote(footnote.to_string());
        self
    }
}

impl<'src> Diagnostic<Source<'src>> {
    /// Calculates the left padding necessary for this diagnostic.
    fn left_padding(&self) -> usize {
        let mut padding = 0;
        for label in &self.labels {
            // find last line of label
            let line_index = self
                .source
                .line_index_at(label.span.end().saturating_sub(1));
            let index_algs = line_index
                .map(|x| f32::log10(x as f32).floor() as usize)
                .unwrap_or(0);
            if index_algs > padding {
                padding = index_algs;
            }
        }

        padding + 1
    }

    fn write_header<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        writeln!(writer, "{}", self.message)?;

        let left_padding = self.left_padding();
        write!(writer, "{:padding$}", "", padding = left_padding)?;
        writeln!(
            writer,
            " {} {}{}{}",
            '@'.style(config.styles.left_column),
            '['.style(config.styles.left_column),
            self.source
                .name()
                .unwrap_or("unknown")
                .style(config.styles.source_name),
            ']'.style(config.styles.left_column)
        )?;
        Ok(())
    }

    fn write_header_compact<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        writeln!(writer, "{}", self.message)?;
        writeln!(
            writer,
            "{} {}{}{}",
            '@'.style(config.styles.left_column),
            '['.style(config.styles.left_column),
            self.source
                .name()
                .unwrap_or("unknown")
                .style(config.styles.source_name),
            "]:".style(config.styles.left_column)
        )?;
        Ok(())
    }

    fn write_body<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        let body_lines = BodyWriter::new(
            writer,
            self.source.clone(),
            config,
            self.left_padding(),
            self.labels.as_slice(),
        );

        body_lines.write()?;

        Ok(())
    }

    fn write_body_compact<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        for label in &self.labels {
            let range = label.line_range(&self.source);
            if range.start + 1 == range.end {
                writeln!(
                    writer,
                    "{} {}{}{}{}{}",
                    config.charset.vertical_bar.style(config.styles.left_column),
                    '['.style(config.styles.left_column),
                    "line ".style(config.styles.source),
                    range.start.style(config.styles.source),
                    "]: ".style(config.styles.left_column),
                    label.message
                )?;
            } else {
                writeln!(
                    writer,
                    "{} {}{}{:?}{}{}",
                    config.charset.vertical_bar.style(config.styles.left_column),
                    '['.style(config.styles.left_column),
                    "lines ".style(config.styles.source),
                    range.style(config.styles.source),
                    "]: ".style(config.styles.left_column),
                    label.message
                )?;
            }
        }

        Ok(())
    }

    fn write_footnotes<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        let left_padding = self.left_padding();

        for footnote in &self.footnotes {
            write!(
                writer,
                "{:padding$} {} ",
                "",
                '>'.style(config.styles.footnote_indicator),
                padding = left_padding
            )?;
            writeln!(writer, "{}", footnote)?;
        }

        Ok(())
    }

    fn write_footnotes_compact<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        for footnote in &self.footnotes {
            write!(writer, "{} ", '>'.style(config.styles.footnote_indicator))?;
            writeln!(writer, "{}", footnote)?;
        }

        Ok(())
    }

    /// Writes this diagnostic to the given [`Write`]r using the specified [`Config`].
    pub fn write_to<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        self.write_header(writer, config)?;
        self.write_body(writer, config)?;
        self.write_footnotes(writer, config)?;

        writeln!(writer)?;
        Ok(())
    }

    /// Writes this diagnostic to the given [`Write`]r using the specified [`Config`]
    /// in compact mode.
    pub fn write_to_compact<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        self.write_header_compact(writer, config)?;
        self.write_body_compact(writer, config)?;
        self.write_footnotes_compact(writer, config)?;

        writeln!(writer)?;
        Ok(())
    }

    /// Writes this diagnostic to `stderr` using the specified [`Config`].
    #[inline]
    pub fn eprint(&self, config: &Config) -> std::io::Result<()> {
        let mut eout = BufWriter::new(std::io::stderr());
        self.write_to(&mut eout, config)?;
        Ok(())
    }

    /// Writes this diagnostic to `stderr` using the specified [`Config`]
    /// in compact mode.
    #[inline]
    pub fn eprint_compact(&self, config: &Config) -> std::io::Result<()> {
        let mut eout = BufWriter::new(std::io::stderr());
        self.write_to_compact(&mut eout, config)?;
        Ok(())
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::test::diagnostic_snapshot;

    #[test]
    fn test_singleline() {
        const SAMPLE: &str = include_str!("../samples/sample2.rs");
        let src = Source::new(SAMPLE, Some("src/lib.rs"));
        let diagnostic = Diagnostic::new("error[E0072]: recursive type `List` has infinite size")
            .with_label(Label::new(53..66u32, ""))
            .with_label(Label::new(83..87u32, "recursive without indirection"))
            .with_footnote("error: could not compile `playground` (lib) due to previous error")
            .with_source(src);

        diagnostic_snapshot!(diagnostic);
    }
}
