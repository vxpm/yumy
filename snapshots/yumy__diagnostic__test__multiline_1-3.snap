---
source: src/diagnostic.rs
expression: clean
---
error[E0277]: `Rc<Mutex<i32>>` cannot be sent between threads safely
@ [src/main.rs]:
│ [line  13]: required by a bound introduced by this call
│ [lines 13..18]: `Rc<Mutex<i32>>` cannot be sent between threads safely
> note: required because it's used within `{closure@src/main.rs:11:36: 11:43}`
> help: within `{closure@src/main.rs:11:36: 11:43}`, the trait `Send` is not implemented for `Rc<Mutex<i32>>`


